### PR TITLE
Issue #504: try merging invoca .rubocop.yml

### PR DESCRIPTION
**Pull request by ColinDKelley**
I made an attempt to address issue #504 :
- Merged in https://github.com/Invoca/style-guide/blob/master/ruby/.rubocop.yml as a start.
- Disabled a few cops and tuned a couple more to better match this repo.
- Ran `rubocop -a`: https://github.com/guard/listen/commit/931106f0a324c82e2a38c83e8944a3d01bbb6010
- Ran `rubocop -A`: https://github.com/guard/listen/commit/8c5fed6b8edbff85b48e9dcaec29b257d7c36635

<details>
  <summary>After the above tuning and auto-correcting, there were 15 remaining offenses:</summary>

```
lib/listen/event/processor.rb:73:9: C: Naming/MemoizedInstanceVariableName: Memoized variable @first_unprocessed_event_time does not match method name _remember_time_of_first_unprocessed_event. Use @_remember_time_of_first_unprocessed_event instead.
        @first_unprocessed_event_time ||= _timestamp
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/listen/options.rb:19:5: C: Style/MissingRespondToMissing: When using method_missing, define respond_to_missing?.
    def method_missing(name, *_) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/listen/thread.rb:18:9: W: Lint/RescueException: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
        rescue Exception => ex ...
        ^^^^^^^^^^^^^^^^^^^^^^
lib/listen/thread.rb:21:9: C: Style/MultilineBlockChain: Avoid multi-line chains of blocks.
        end.tap do |thread|
        ^^^^^^^
lib/listen/thread.rb:28:26: C: Naming/MethodParameterName: Method parameter must be at least 3 characters long.
      def _log_exception(ex, thread_name, caller_stack)
                         ^^
lib/listen/thread.rb:34:34: C: Naming/MethodParameterName: Method parameter must be at least 3 characters long.
      def _exception_with_causes(ex)
                                 ^^
listen.gemspec:30:31: C: Gemspec/RequiredRubyVersion: required_ruby_version (2.2, declared in listen.gemspec) and TargetRubyVersion (2.6, which may be specified in .rubocop.yml) should be equal.
  gem.required_ruby_version = '>= 2.2.7'
                              ^^^^^^^^^^
spec/acceptance/listen_spec.rb:20:48: W: Lint/LiteralAsCondition: Literal 'both' appeared as a condition.
      case ENV['TEST_LISTEN_ADAPTER_MODES'] || 'both'
                                               ^^^^^^
spec/lib/listen/directory_spec.rb:16:21: C: Naming/MethodParameterName: Method parameter must be at least 3 characters long.
  def fake_children(ex, dir, *args, &block)
                    ^^
spec/lib/listen/thread_spec.rb:32:151: C: Layout/LineLength: Line is too long. [177/150]
                                 with(/Exception rescued in listen-worker_thread:\nArgumentError: boom!\n.*\/listen\/thread\.rb.*--- Thread.new ---.*\/listen\/thread_spec\.rb/m)
                                                                                                                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
spec/lib/listen/thread_spec.rb:54:34: C: Layout/DotPosition: Place the . on the previous line, together with the method call receiver.
                                 .with(/RuntimeError: nested outer\n--- Caused by: ---\nRuntimeError: nested inner\n--- Caused by: ---\nArgumentError: boom!/)
                                 ^
spec/lib/listen/thread_spec.rb:54:151: C: Layout/LineLength: Line is too long. [158/150]
                                 .with(/RuntimeError: nested outer\n--- Caused by: ---\nRuntimeError: nested inner\n--- Caused by: ---\nArgumentError: boom!/)
                                                                                                                                                      ^^^^^^^^
spec/support/acceptance_helper.rb:46:5: C: Security/Open: The use of Kernel#open is a serious security risk.
    open(path, 'a') { |f| f.write('foo') }
    ^^^^
spec/support/acceptance_helper.rb:59:5: C: Security/Open: The use of Kernel#open is a serious security risk.
    open(path, 'w') { |f| f.write('foo') }
    ^^^^
spec/support/acceptance_helper.rb:181:3: W: Lint/DuplicateMethods: Method ListenerWrapper#changes is defined at both spec/support/acceptance_helper.rb:146 and spec/support/acceptance_helper.rb:181.
  def changes
  ^^^^^^^^^^^
```

</details>

<details>
  <summary>After fixing the easy ones, there are 5 offenses left:</summary>

```
lib/listen/options.rb:19:5: C: Style/MissingRespondToMissing: When using method_missing, define respond_to_missing?.
    def method_missing(name, *_) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/listen/thread.rb:18:9: W: Lint/RescueException: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
        rescue Exception => exception ...
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
listen.gemspec:30:31: C: Gemspec/RequiredRubyVersion: required_ruby_version (2.2, declared in listen.gemspec) and TargetRubyVersion (2.6, which may be specified in .rubocop.yml) should be equal.
  gem.required_ruby_version = '>= 2.2.7'
                              ^^^^^^^^^^
spec/support/acceptance_helper.rb:46:5: C: Security/Open: The use of Kernel#open is a serious security risk.
    open(path, 'a') { |f| f.write('foo') }
    ^^^^
spec/support/acceptance_helper.rb:59:5: C: Security/Open: The use of Kernel#open is a serious security risk.
    open(path, 'w') { |f| f.write('foo') }
    ^^^^

65 files inspected, 5 offenses detected
```

</detail>

### CODE REVIEW

**Comment by ColinDKelley** [comment]
@KapilSachdev What do you think?

File: .rubocop.yml [hunk 1]
```diff
@@ -7,13 +7,190 @@ AllCops:
 
 # Files you want to exclude
 AllCops:
-  TargetRubyVersion: 2.3
+  TargetRubyVersion: 2.6
```
Comment (KapilSachdev) [inline .rubocop.yml:10]: Any reason for not targeting 2.7?

File: Gemfile [hunk 2]
```diff
@@ -34,6 +34,6 @@ group :development do
   gem 'netrc', require: false
   gem 'octokit', require: false
   gem 'pry-rescue'
-  gem 'rubocop', '~> 0.49.0' # TODO: should match Gemfile HoundCi
+  gem 'rubocop', '0.91.0'
```
Comment (KapilSachdev) [inline Gemfile:37]: We can update to `1.2.0`, I think there will be not much changes as I have checked in my local machine.

Comment (ColinDKelley) [reply Gemfile:37]: Currently, the [highest supported by hound](http://help.houndci.com/en/articles/2461415-supported-linters) is 0.91.0.

File: .rubocop.yml (hunk 1 above)
Comment (ColinDKelley) [reply .rubocop.yml]: Done.

File: lib/listen/thread.rb [hunk 3]
```diff
@@ -14,12 +14,10 @@ def new(name)
...
         caller_stack = caller
         ::Thread.new do
-          begin
-            yield
-          rescue Exception => ex
-            _log_exception(ex, thread_name, caller_stack)
-            nil
-          end
+          yield
+        rescue Exception => ex
```
[the remark this answers is not in the document]
Comment (ColinDKelley) [reply lib/listen/thread.rb]: Nope. `Exception` is intentional.

File: lib/listen/adapter.rb [hunk 4]
```diff
@@ -24,7 +24,7 @@ def select(options = {})
         Polling
       rescue
         Listen.logger.warn format('Adapter: failed: %s:%s', $ERROR_POSITION.inspect,
-                           $ERROR_POSITION * "\n")
+                                  $ERROR_POSITION * "\n")
```
Comment (KapilSachdev) [inline lib/listen/adapter.rb:27]: Style are personal preferences, but giving more spaces looks unnecessary to me.

File: lib/listen/thread.rb [hunk 5]
```diff
@@ -11,34 +11,32 @@ class << self
...
-          rescue Exception => ex
-            _log_exception(ex, thread_name, caller_stack)
-            nil
-          end
-        end.tap do |thread|
-          thread.name = thread_name
+
+        thread = ::Thread.new do
+          yield
+        rescue Exception => exception
```
[the remark this answers is not in the document]
Comment (ColinDKelley) [reply lib/listen/thread.rb:18]: Nope. `Exception` is intentional.

File: lib/listen/adapter/bsd.rb [hunk 6]
```diff
@@ -7,7 +7,7 @@
 module Listen
   module Adapter
     class BSD < Base
-      OS_REGEXP = /bsd|dragonfly/i
+      OS_REGEXP = /bsd|dragonfly/i.freeze
```
Comment (KapilSachdev) [inline lib/listen/adapter/bsd.rb:10]: This file can not use `frozen_string_literal`?

File: lib/listen/adapter.rb (hunk 4 above)
Comment (ColinDKelley) [reply lib/listen/adapter.rb:27]: `rubocop` did that. I've stopped trying to predict its preference for spacing.

File: spec/acceptance/listen_spec.rb [hunk 7]
```diff
@@ -144,8 +142,7 @@
             it 'listens to file move' do
               expected = { modified: [],
                            added: %w(file.rb),
-                           removed: %w(dir/file.rb)
-              }
+                           removed: %w(dir/file.rb) }
```
Comment (KapilSachdev) [inline spec/acceptance/listen_spec.rb:145]: The same extra indentation could be avoided here as well. Though its upto the maintainer's preference.
```ruby
expected =
  { 
    modified: [],
    added: %w(file.rb),
    removed: %w(dir/file.rb)
  }
```

File: lib/listen/adapter/bsd.rb (hunk 6 above)
Comment (ColinDKelley) [reply lib/listen/adapter/bsd.rb:10]: The magic comment is there, but doesn't affect regular expressions. `rubocop` knows that regular expression literals are mutable, although I don't understand how.
https://github.com/rubocop-hq/rubocop/issues/6331

File: spec/lib/listen/backend_spec.rb [hunk 8]
```diff
@@ -33,33 +33,33 @@
     stub_const('Listen::Adapter::Config', adapter_config_class)
 
     allow(adapter_config_class).to receive(:new).
-      with([dir1], queue, silencer, adapter_options).
```
Comment (KapilSachdev) [inline spec/lib/listen/backend_spec.rb:36]: The current indentation looks better to me. This is one of the side effects as line length offence gets into action sometimes while indenting.

File: spec/acceptance/listen_spec.rb (hunk 7 above)
Comment (ColinDKelley) [reply spec/acceptance/listen_spec.rb:145]: Please propose a change to the `rubocop` rules so it tolerates the formatting you like.

File: spec/lib/listen/thread_spec.rb [hunk 9]
```diff
@@ -22,14 +22,14 @@
...
+      expect(Listen.logger).to receive(:error).
+                                 with(/Exception rescued in listen-worker_thread:\nArgumentError: boom!\n.*\/listen\/thread_spec\.rb/)
       subject.join
     end
 
     it "rescues and logs backtrace + exception backtrace" do
-      expect(Listen.logger).to receive(:error)
-        .with(/Exception rescued in listen-worker_thread:\nArgumentError: boom!\n.*\/listen\/thread\.rb.*--- Thread.new ---.*\/listen\/thread_spec\.rb/m)
+      expect(Listen.logger).to receive(:error).
+                                 with(/Exception rescued in listen-worker_thread:\nArgumentError: boom!\n.*\/listen\/thread\.rb.*--- Thread.new ---.*\/listen\/thread_spec\.rb/m)
```
Comment (KapilSachdev) [inline spec/lib/listen/thread_spec.rb]: Looks like we encountered the side effect of extra indentation.

File: spec/lib/listen/backend_spec.rb (hunk 8 above)
Comment (ColinDKelley) [reply spec/lib/listen/backend_spec.rb:36]: Same comment.

File: spec/support/acceptance_helper.rb [hunk 10]
```diff
@@ -162,20 +162,20 @@ def initialize(callback, paths, *args)
...
+                    _add_changes(:added, added, changes)
+                    _add_changes(:removed, removed, changes)
+
+                    unless callback == :track_changes
+                      callback.call(modified, added, removed)
+                    end
+                  end
+                else
+                  Listen.send(*args)
+                end
```
Comment (KapilSachdev) [inline spec/support/acceptance_helper.rb:178]: Adding the assignment in the next line could help in less indentation spaces.  
```suggestion
    @listener =
  if callback
    Listen.send(*args) do |modified, added, removed|
      # Add changes to trigger frozen Hash error, making sure lag is enough
      _add_changes(:modified, modified, changes)
      _add_changes(:added, added, changes)
      _add_changes(:removed, removed, changes)

      unless callback == :track_changes
        callback.call(modified, added, removed)
      end
    end
  else
    Listen.send(*args)
  end
```

File: vendor/hound/config/style_guides/ruby.yml [hunk 11]
```diff
@@ -60,7 +60,7 @@ Naming/PredicateName:
   - is_
```
Comment (KapilSachdev) [inline vendor/hound/config/style_guides/ruby.yml:60]: Is this file necessary for hound to work?
Actually I had removed the `vendor/hound/config/style_guides/ruby.yml` when i started making changes.

File: Gemfile (hunk 2 above)
Comment (KapilSachdev) [reply Gemfile:37]: We can use something else than hound in that case? Github actions like [lint-action](https://github.com/marketplace/actions/lint-action).

File: lib/listen/adapter.rb (hunk 4 above)
Comment (KapilSachdev) [reply lib/listen/adapter.rb:27]: Okay.

File: lib/listen/adapter/bsd.rb (hunk 6 above)
Comment (KapilSachdev) [reply lib/listen/adapter/bsd.rb:10]: I looked at the issue you mentioned, and I think `rubocop` differentiates while frozing literal type, so the change is required. But I will review if it actually is what i'm thinking.

File: spec/acceptance/listen_spec.rb (hunk 7 above)
Comment (KapilSachdev) [reply spec/acceptance/listen_spec.rb:145]: I'm not sure why `rubocop -a` does not did that but `Layout/MultilineAssignmentLayout` defaults to new line.
Maybe running `rubocop -a --only Layout/MultilineAssignmentLayout` might work?

File: spec/lib/listen/backend_spec.rb (hunk 8 above)
Comment (KapilSachdev) [reply spec/lib/listen/backend_spec.rb:36]: This should work.
```ruby
Layout/MultilineMethodCallIndentation:
  EnforcedStyle: indented
```

File: lib/listen/thread.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (KapilSachdev) [reply lib/listen/thread.rb]: We should add `# rubocop:disable Lint/RescueException` for intentional offences

```ruby
rescue Exception => ex # rubocop:disable Lint/RescueException
```

File: Gemfile (hunk 2 above)
Comment (ColinDKelley) [reply Gemfile:37]: @KapilSachdev That sounds reasonable. So that we don't bog this Issue down, can you file a separate issue for that?

File: vendor/hound/config/style_guides/ruby.yml (hunk 11 above)
Comment (ColinDKelley) [reply vendor/hound/config/style_guides/ruby.yml:60]: I'm not sure, but I can try removing it.

File: spec/lib/listen/backend_spec.rb (hunk 8 above)
Comment (ColinDKelley) [reply spec/lib/listen/backend_spec.rb]: Changed.

File: vendor/hound/config/style_guides/ruby.yml (hunk 11 above)
Comment (ColinDKelley) [reply vendor/hound/config/style_guides/ruby.yml]: Merged and removed.

**Comment by ColinDKelley** [comment]
@KapilSachdev I merged up to latest master and applied all the changes in discussion here.
```
$ rubocop
Inspecting 65 files
.................................................................

65 files inspected, no offenses detected
```

File: spec/support/acceptance_helper.rb (hunk 10 above)
Comment (ioquatix) [reply spec/support/acceptance_helper.rb:174]: I also agree, this alignment change is a mess, and it should probably just be

```
@listener = if callback
  Listen.send(...)
else
  ...
end
```

Comment (ColinDKelley) [reply spec/support/acceptance_helper.rb]: Rubocop wouldn't tolerate that formatting. I put 10 minutes into finding an option that would, but I didn't find it. Instead, I tried the `if` on the next line and it was ok with that. Of course then it complained that the method was too long! So I moved an `unless` from leading to trailing. 🤦 

https://github.com/guard/listen/pull/508/commits/745d0e35a433b5fe0bd8064ef6bdf0e5fc28e36b

File: Gemfile (hunk 2 above)
Comment (KapilSachdev) [reply Gemfile:37]: The time i commented about `lint-action`, i was unaware of one issue that they have i.e. annotations don't work when PR is created from a fork. Which is where most PRs come from. Though there is a new `pull_request_target` event for that, but it still isn't working (from what i had tried).
`shoulda-matcher` gem also recently reverted from `lint-action` (which initially uses hound) and has fallen back to non-annotation based `rubocop` linting (like what rails uses).

I will create a issue for `rubocop` linting based on github actions.

File: spec/support/acceptance_helper.rb (hunk 10 above)
Comment (KapilSachdev) [reply spec/support/acceptance_helper.rb]: I think below should help. However as you have used next line, you won't need this then. Next lines looks good to me.
```yml
Layout/EndAlignment:
  EnforcedStyleAlignWith: variable
```

Comment (ColinDKelley) [reply spec/support/acceptance_helper.rb]: Thanks @KapilSachdev. I made that `EnforcedStyleAlignWith: variable ` change and put the `@listener =` code back the way it was, then ran `rubocop -a -A`.  https://github.com/guard/listen/pull/508/commits/e1045d21d8cef8dd1752897f9d037a21a4e1f652
